File: plugins/apbdoor/apbdoor.go
package apbdoor

import (
	"bytes"
	"io/ioutil"
	"net/http"
	"regexp"
	"strings"

	"github.com/abourget/slick"
	"github.com/fsr/matthias/util/random"
)

const (
	doorBroken     = "yes"
	doorFunctional = "no"
	doorUnknown    = "maybe"
)

type apbdoor struct{}

func (apbdoor *apbdoor) String() string {
	return `!türstatus/!tuerstatus - Ist die Eingangstür vom APB aktuell im Eimer?
!tuer (ist) (kaputt/ganz/weg) - Setzt die Eingangstür vom APB auf den jeweiligen Status`
}

func init() {
	slick.RegisterPlugin(&apbdoor{})
}

// InitPlugin ...
func (apbdoor *apbdoor) InitPlugin(bot *slick.Bot) {
	bot.Listen(&slick.Listener{
		Matches:            regexp.MustCompile("!t(?:ü|ue)rstatus"),
		MessageHandlerFunc: apbdoor.checkDoorHandler,
	})
	bot.Listen(&slick.Listener{
		Matches:            regexp.MustCompile("!t(?:ü|ue)r(?: ist )?(.+)"),
		MessageHandlerFunc: apbdoor.setDoorHandler,
	})
}

func (apbdoor *apbdoor) checkDoorHandler(listen *slick.Listener, msg *slick.Message) {
	doorState, err := checkDoorState()
	if err != nil {
		msg.Reply(random.StringFromList(maybeMsgs))
	}

	switch doorState {
	case doorBroken:
		msg.Reply(random.StringFromList(yesMsgs))
	case doorFunctional:
		msg.Reply(random.StringFromList(noMsgs))
	default:
		msg.Reply(random.StringFromList(maybeMsgs))
	}
}

func (apbdoor *apbdoor) setDoorHandler(listen *slick.Listener, msg *slick.Message) {
	state := strings.ToLower(msg.Match[1])
	switch state {
	case "kaputt", "broken", "im eimer":
		setDoorState(doorBroken)
		msg.Reply("Orr ne, schon wieder?!")
	case "ganz", "wieder ganz", "funktional":
		setDoorState(doorFunctional)
		msg.Reply(random.StringFromList(partyGifs))
	case "weg", "unbekannt":
		setDoorState(doorUnknown)
		msg.Reply("Ähm.... Ahja?")
	}
}

func checkDoorState() (string, error) {
	resp, err := http.Get("http://tuer.fsrleaks.de")
	defer resp.Body.Close()
	if err != nil {
		return doorUnknown, err
	}
	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return doorUnknown, err
	}

	if bytes.Contains(body, []byte("Ja")) {
		return doorBroken, nil
	} else if bytes.Contains(body, []byte("Nein")) {
		return doorFunctional, nil
	} else {
		return doorUnknown, nil
	}
}

func setDoorState(state string) {
	url := "http://door.fsrleaks.de/set.php?" + state
	http.Get(url)
}

var partyGifs = []string{
	"http://i.giphy.com/6nuiJjOOQBBn2.gif",
	"http://i.giphy.com/EktbegF3J8QIo.gif",
	"http://i.giphy.com/YTbZzCkRQCEJa.gif",
	"http://i.giphy.com/3rgXBQIDHkFNniTNRu.gif",
	"http://i.giphy.com/s2qXK8wAvkHTO.gif",
}

var yesMsgs = []string{
	"Jop, Tür ist im Eimer.",
	"Tür ist 'putt.",
	"Rate mal...",
	"Alles im Arsch, Normalzustand halt.",
	"Computer sagt: Tür ist hin.",
	"Tür pass auf!! Du hast ne Scheibe verloren!",
	"Techniker ist selbstverständlich bereits informiert.",
}

var noMsgs = []string{
	"Die Tür ist... ganz?!",
	"Alles im grünen Bereich.",
	"Sie ist ganz! Also... Zumindest gerade eben. Vermutlich schon nicht mehr.",
	"Rufe Fr. Kapplusch an... Nop, scheint alles gut zu sein. Beeindruckend.",
	"Hab' eben nachgesehen und... Ausnahmezustand!",
}

var maybeMsgs = []string{
	"Ich hab keine Ahnung ¯\\_(ツ)_/¯",
	"Sorry, musst du dieses Mal selber nachschauen.",
	"Quizás, señor/a.",
	"Sorry, no clue.",
	"Schrödingers Tür?",
	"Lässt sich aus diesem Blickwinkel schlecht beurteilen.",
}

File: bot/Apbdoor.hs
-- Description:
--   Der Techniker ist informiert.
-- Dependencies:
--   None

-- Configuration:
--   None

-- Commands:
--   hubot türstatus - hubot schaut nach, wie es um die Tür des APB steht.
--   hubot ist die tür kaputt? - hubot schaut nach, wie es um die Tür des APB steht.
--   hubot glasschaden - teile hubot mit, dass die Tür mal wieder im Eimer ist.
--   hubot rate mal, was wieder kaputt ist - teile hubot mit, dass die Tür mal wieder im Eimer ist.
--   hubot techniker ist informiert - teile hubot mit, dass die Tür mal wieder im Eimer ist.
--   hubot tür ist wieder ganz - teile hubot mit, dass die Tür wieder ganz ist
--   hubot tür ist weg - teile hubot mit, dass die Tür sich in einem unbekannten Status befindet.

-- Author:
--   kiliankoe
{-# LANGUAGE MultiWayIf #-}
module Apbdoor (script) where


import Marvin.Prelude
import Network.Wreq
import Control.Lens


script :: IsAdapter a => ScriptInit a
script = defineScript "apbdoor" $ do
  respond (r [CaseInsensitive] "türstatus|tuerstatus|ist die tür kaputt\\?|ist die tuer kaputt\\?") $
    checkDoor

  respond (r [CaseInsensitive] "glasschaden|rate mal, was wieder kaputt ist|techniker ist informiert") $ do
    setDoor "yes"
    send "Orr ne, schon wieder?!"

  respond (r [CaseInsensitive] "tür ist wieder ganz|tuer ist wieder ganz") $ do
    setDoor "no"
    send "/giphy party"

  respond (r [CaseInsensitive] "tür ist weg|tuer ist weg") $ do
    setDoor "maybe"
    send "Ähm... Ahja?"


checkDoor :: IsAdapter a => BotReacting a MessageReactionData ()
checkDoor = do
  r <- liftIO $ get "http://tuer.fsrleaks.de"
  
  let body = r^.responseBody
  -- TODO Add random
  if 
    | "Ja" `isInfixOf` body -> randomFrom yesMsgs >>= send
    | "Nein" `isInfixOf` body -> randomFrom noMsgs >>= send
    | otherwise -> randomFrom maybeMsgs >>= send


setDoor state = liftIO $ get (unpack $ format "http://door.fsrleaks.de/set.php?{}" [state :: Text])

yesMsgs = 
  [ "Jop, Tür ist im Eimer."
  , "Tür ist 'putt."
  , "Rate mal..."
  , "Alles im Arsch, Normalzustand halt."
  , "Computer sagt: Tür ist hin."
  , "Tür pass auf!! Du hast ne Scheibe verloren!"
  , "Techniker ist selbstverständlich bereits informiert."
  ]

noMsgs = 
  [ "Die Tür ist... ganz?!"
  , "Alles im grünen Bereich."
  , "Sie ist ganz! Also... Zumindest gerade eben. Vermutlich schon nicht mehr."
  , "Rufe Fr. Kapplusch an... Nop, scheint alles gut zu sein. Beeindruckend."
  , "Hab' eben nachgesehen und... Ausnahmezustand!"
  ]

maybeMsgs = 
  [ "Ich hab keine Ahnung ¯\\_(ツ)_/¯"
  , "Sorry, musst du dieses Mal selber nachschauen."
  , "Quizás, señor/a."
  , "Sorry, no clue."
  , "Schrödingers Tür?"
  , "Lässt sich aus diesem Blickwinkel schlecht beurteilen."
  ]
